lab8/t8.c

```diff
@@ -0,0 +1,268 @@
+
+    for (int i = 0; i < m; i++) {
+        printf("matrix [%d]: \n", i);
+        for (int j = 0; j < n; j++) {
+            scanf("%lf", &matrix[i][j]); 
+        }
+    }
+
+}
+
+void task4(){
+
+    int m, n;
+    printf("Enter (m n): ");  
+    scanf("%d %d", &m, &n);
+
+    double matrix[25][25]; 
+    input_matrix4(matrix, m, n);
+
+    printf("Entered matrix: \n");
+    for (int i = 0; i < m; i++) {
+        for (int j = 0; j < n; j++) {
+            printf("%lf ", matrix[i][j]);
+        }
+        printf("\n");
+    }
+
+}
+
+void input_matrix(double matrix[MAX_SIZE][MAX_SIZE], int m, int n) {
+    for (int i = 0; i < m; i++){
+        for (int j = 0; j < n; j++){
+            printf("Enter value of matrix [%d][%d]: ", i, j);
+            scanf("%lf", &matrix[i][j]); 
+        }
+    }
+}
+
+void output_matrix(double matrix[MAX_SIZE][MAX_SIZE], int m, int n) {
+    for (int i = 0; i < m; i++){
+        for (int j = 0; j < n; j++){
+            printf("%.2lf ", matrix[i][j]);
+        }
+        printf("\n");
+    }
+}
+
+void transpose_matrix(double matrix[MAX_SIZE][MAX_SIZE], double transposed[MAX_SIZE][MAX_SIZE], int m, int n) {
+    for (int i = 0; i < m; i++){
+        for (int j = 0; j < n; j++){
+            transposed[i][j] = matrix[j][i];
+        }
+    }
+}
+
+void task5() {
+    
+    int M, N;
+    printf("M, N: ");
+    scanf("%d %d", &M, &N);
+    
+    double matrix[MAX_SIZE][MAX_SIZE], transposed[MAX_SIZE][MAX_SIZE];
+    
+    printf("Enter matrix elements:\n");
+    input_matrix(matrix, M, N);
+    
+    printf("Original matrix:\n");
+    output_matrix(matrix, M, N);
+
+    transpose_matrix(matrix, transposed, M, N);
+
+    printf("Transposed matrix:\n");
+    output_matrix(transposed, M, N);
+
+}
+
+double calculate_sum(double A[MAX_SIZE][MAX_SIZE], int N, int M, int k){
+
+    double sum=0.0;
+    int count=0; 
+
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < M; j++) {
+            if (i - j == k) {
+                sum += A[i][j];
+                count++; 
+            }
+        }
+    }
+
+    if (count == 0){
+        return 0;
+    }
+    else{
+        return sum;
+    }
+
+}
+
+void task6(){
+
+    int N, M, k;
+    printf("N, M (<100) = ");
+    scanf("%d %d", &N, &M);
+    double A[MAX_SIZE][MAX_SIZE];
+    
+    input_matrix(A, M, N);
+
+    printf("Enter value of k: ");
+    scanf("%d", &k);
+
+    printf("The sum of elements where i - j = %d is: %.2lf \n", k, calculate_sum(A, N, M, k));
+
+}
+
+double determinant(double matrix[MAX_SIZE][MAX_SIZE], int n){
+
+    if (n == 1) return matrix[0][0];
+
+    double det = 0;
+
+    // цикл по стовпцям
+    for (int j = 0; j < n; j++) { 
+        int sign = (j % 2 == 0) ? 1 : -1; 
+        double submatrix[MAX_SIZE][MAX_SIZE];
+        for (int i = 1; i < n; i++) { 
+            for (int k = 0, l = 0; l < n; l++) {
+                if (l == j) continue; 
+                submatrix[i - 1][k] = matrix[i][l];
```

Comment:
Not sure this works, could You show me the tests?

Reply:
> Not sure this works, could You show me the tests?

I reworked the algorithm for finding the determinant, now convert the matrix into an upper triangular one and count the diagonal == determinant (now it works properly)
+ wrote the tests